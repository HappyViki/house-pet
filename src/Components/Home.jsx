import React from 'react'

class Home extends React.Component {
  constructor (props) {
    super(props)
    this.state = {
      pets: null,
      currentPet: 0,
      current: {
        id: 0,
        src: '#',
        distance: null,
        url: ''
      }
    }
  }

  componentDidMount () {
    let pets = window.localStorage.getItem('pets')
    if (pets === null) {
      pets = this.fetchPets()
      console.log('Loaded new pets', pets)
    } else {
      pets = JSON.parse(pets)
      console.log('Loaded old pets', pets)
    }
    this.setState({ pets: pets }, () => {
      this.setPet(0)
    })
  }

  fetchPets () {
    fetch('/api')
      .then(response => {
        return response.json()
      })
      .then(json => {
        const petsWithPics = json.filter((pet) => {
          if (pet.photos.length > 0) {
            return pet
          }
        })
        const pets = JSON.stringify(petsWithPics)
        window.localStorage.setItem('pets', pets)
        return petsWithPics
      })
  }

  nextPet () {
    console.log('Next pet!')
    let newCurrentPet = this.state.currentPet + 1
    if (newCurrentPet > this.state.pets.length - 1) {
      newCurrentPet = 0
    }
    this.setState({ currentPet: newCurrentPet })
    this.setPet(newCurrentPet)
  }

  setPet (currentPet) {
    if (this.state.pets !== null && this.state.pets !== undefined) {
      let newSrc = '#'
      if (this.state.pets[currentPet].photos.length > 0) {
        newSrc = this.state.pets[currentPet].photos[0].medium
      }
      this.setState({
        current: {
          id: this.state.pets[currentPet].id,
          src: newSrc,
          distance: this.state.pets[currentPet].distance,
          url: this.state.pets[currentPet].url
        }
      })
    }
  }

  savePet () {
    const storage = window.localStorage
    const newPetObject = {
      id: this.state.current.id,
      src: this.state.current.src,
      distance: this.state.current.distance,
      url: this.state.current.url
    }
    console.log('savePet', newPetObject)
    let myPets = storage.getItem('myPets') ? JSON.parse(storage.getItem('myPets')) : {}
    if (!myPets[newPetObject.id]) {
      myPets[newPetObject.id] = newPetObject
      myPets = JSON.stringify(myPets)
      storage.setItem('myPets', myPets)
    }
    // window.localStorage.getItem('myPets')
  }

  render () {
    let distanceMessage = "Location wasn't provided."
    if (this.state.current.distance !== null) {
      distanceMessage = `You are ${this.state.current.distance} miles away!`
    }

    return (
      <div className='card'>
        <figure className='image image-container'>
          <img src={this.state.current.src} />
        </figure>
        <div className='content'>
          {distanceMessage}
        </div>
        <div className='buttons'>
          <button className='button is-warning' onClick={() => this.nextPet()}>
            Noop
          </button>
          <button className='button is-success' onClick={() => this.savePet()}>
            Boop
          </button>
        </div>
      </div>
    )
  }
}

export default Home
