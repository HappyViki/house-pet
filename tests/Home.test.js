import React from 'react'
import Home from '../src/Components/Home.jsx'
import renderer from 'react-test-renderer'

it('renders correctly', () => {
  global.fetch = jest.fn(() => Promise.resolve(null).catch())
  const tree = renderer
    .create(<Home />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})
