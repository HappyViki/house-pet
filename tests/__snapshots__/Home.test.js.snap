// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="home"
>
  <img
    className="pic"
    src="#"
  />
  <div
    className="info"
  >
    <b>
      Distance from your location:
    </b>
    <br />
    Location wasn't provided.
  </div>
  <button
    onClick={[Function]}
  >
    Noop
  </button>
  <button
    onClick={[Function]}
  >
    Boop
  </button>
  <div>
    <b>
      Dev Tools:
    </b>
    <button
      onClick={[Function]}
    >
      Fetch Pets
    </button>
    <button
      onClick={[Function]}
    >
      Clear Cache
    </button>
  </div>
</div>
`;
